### PR TITLE
Add support for IN and OUT parameters in PLAY, CALL and SEEK commands

### PR DESCRIPTION
**Pull request by dimitry-ishenko**
This patch adds support for the following `PLAY` command syntax:

    PLAY 1-10 folder/clip IN 10 LENGTH 50
    PLAY 1-10 folder/clip IN 10 OUT 60 (equivalent to the above)

(Use of `SEEK` instead of `IN` is still supported.)

Following `SEEK` command syntax is now supported:

    CALL 1-10 SEEK 123 (seek to absolute frame 123)
    CALL 1-10 SEEK REL 10 (seek 10 frames forward from current position)
    CALL 1-10 SEEK REL -50 (seek 50 frames back from current position)
    CALL 1-10 SEEK IN (seek to IN point)
    CALL 1-10 SEEK IN 10 (seek to 10 frames after IN point)
    CALL 1-10 SEEK IN -50 (seek to 50 frames before IN point)

Additionally, `OUT` and `END` can be specified in place of `IN` to seek relative to the OUT point or the end of the file respectively.

Following `CALL` command syntax is now supported:

    CALL 1-10 IN 10 (set IN point to 10 frames)
    CALL 1-10 START 10 (still supported; equivalent to the above)
    CALL 1-10 OUT 60 (set OUT point to 60 frames)
    CALL 1-10 LENGTH 50 (set OUT point to IN + 50 frames)

As I've mentioned in Issue #499 the behaviour of the `LENGTH` parameter is now modified to be no longer affected by seeking during playback. In other words, when the `LENGTH` command is issued, it is used to calculate the OUT point relative to the current IN point. (NB: If the IN point is modified afterwards, the OUT point is not adjusted to keep the same length.)

### CODE REVIEW

File: modules/ffmpeg/producer/ffmpeg_producer.cpp [hunk 1]
```diff
@@ -313,12 +316,12 @@ struct ffmpeg_producer : public core::frame_producer_base
 			                   // therefore no seeking should be necessary for the first frame.
 		{
 			input_.seek(file_position > 1 ? file_position - 2: file_position).get();
-            boost::this_thread::sleep_for(boost::chrono::milliseconds(40));
+			std::this_thread::sleep_for(std::chrono::milliseconds(40));
```
Comment (HellGore) [inline modules/ffmpeg/producer/ffmpeg_producer.cpp]: Why have you changed this? This removes support for `boost::thread::interrupt()`.

File: modules/ffmpeg/producer/ffmpeg_producer.cpp [hunk 2]
```diff
@@ -405,74 +408,77 @@ struct ffmpeg_producer : public core::frame_producer_base
...
+				else if(boost::iequals(value, L"end"))
+					seek = nb_frames;
+				else
+					seek = boost::lexical_cast<int64_t>(value);
 
-			if(value < 0)
-				value = 0;
-			else if(value >= total)
-				value = total - 1;
+				if(params.size() > 2)
```
Comment (HellGore) [inline modules/ffmpeg/producer/ffmpeg_producer.cpp]: Is it intended to be optional? If no, then it is better to let the out of range exception propagate here as it is handled by https://github.com/CasparCG/Server/blob/2.1.0/protocol/amcp/AMCPCommandQueue.cpp#L129 instead

File: modules/ffmpeg/producer/ffmpeg_producer.cpp (hunk 1 above)
Comment (dimitry-ishenko) [inline modules/ffmpeg/producer/ffmpeg_producer.cpp]: OK fixed

File: modules/ffmpeg/producer/ffmpeg_producer.cpp (hunk 2 above)
Comment (dimitry-ishenko) [inline modules/ffmpeg/producer/ffmpeg_producer.cpp]: Yes, I made it optional by intent. My idea was that you can simply do `CALL 1-10 SEEK OUT` to seek e.g. to OUT point.

File: modules/ffmpeg/producer/ffmpeg_producer.cpp [hunk 3]
```diff
@@ -405,74 +408,77 @@ struct ffmpeg_producer : public core::frame_producer_base
...
+			std::wstring cmd = params.at(0), value;
+			if(params.size() > 1) value = params.at(1);
 
-			if(boost::iequals(whence, L"REL"))
-				value = file_frame_number() + value;
-			else if(boost::iequals(whence, L"END"))
-				value = total - value;
+			if(boost::iequals(cmd, L"loop"))
+			{
+				if(value.size())
```
Comment (HellGore) [inline modules/ffmpeg/producer/ffmpeg_producer.cpp]: Use `!value.empty()` instead, it better demonstrates intent.

File: modules/ffmpeg/producer/ffmpeg_producer.cpp [hunk 4]
```diff
@@ -405,74 +408,77 @@ struct ffmpeg_producer : public core::frame_producer_base
...
 	{
-		uint32_t file_nb_frames = 0;
-		file_nb_frames = std::max(file_nb_frames, video_decoder_ ? video_decoder_->nb_frames() : 0);
-		return file_nb_frames;
+		return video_decoder_ ? video_decoder_->nb_frames() : 0;
 	}
 
 	std::future<std::wstring> call(const std::vector<std::wstring>& params) override
 	{
-		static const boost::wregex loop_exp(LR"(LOOP\s*(?<VALUE>\d?)?)", boost::regex::icase);
```
Comment (HellGore) [inline modules/ffmpeg/producer/ffmpeg_producer.cpp]: I am not too keen on using regexp either so I welcome this change.

File: modules/ffmpeg/producer/input/input.cpp [hunk 5]
```diff
@@ -364,6 +358,8 @@ struct input::implementation : boost::noncopyable
 			static_cast<int64_t>((target / fps * stream->time_base.den) / stream->time_base.num),
 			std::numeric_limits<int64_t>::max(),
 			0), print());
+		
+		file_frame_number_ = target;
```
Comment (HellGore) [inline modules/ffmpeg/producer/input/input.cpp]: Good catch

File: modules/ffmpeg/producer/ffmpeg_producer.cpp (hunk 3 above)
Comment (dimitry-ishenko) [inline modules/ffmpeg/producer/ffmpeg_producer.cpp]: OK, makes sense

File: modules/ffmpeg/producer/ffmpeg_producer.cpp [hunk 6]
```diff
@@ -405,74 +408,77 @@ struct ffmpeg_producer : public core::frame_producer_base
...
-		boost::wsmatch what;
-		if(boost::regex_match(param, what, loop_exp))
-		{
-			auto value = what["VALUE"].str();
-			if (!value.empty())
-				input_.loop(boost::lexical_cast<bool>(value));
-			result = boost::lexical_cast<std::wstring>(input_.loop());
-		}
-		else if(boost::regex_match(param, what, seek_exp))
+		if(params.size())
```
Comment (dimitry-ishenko) [inline modules/ffmpeg/producer/ffmpeg_producer.cpp]: Tracking

File: modules/ffmpeg/producer/ffmpeg_producer.cpp [hunk 7]
```diff
@@ -405,74 +408,77 @@ struct ffmpeg_producer : public core::frame_producer_base
...
 		{
-			auto value = boost::lexical_cast<int64_t>(what["VALUE"].str());
-			auto whence = what["WHENCE"].str();
-			auto total = file_nb_frames();
+			std::wstring cmd = params.at(0), value;
+			if(params.size() > 1) value = params.at(1);
```
Comment (HellGore) [inline modules/ffmpeg/producer/ffmpeg_producer.cpp]: I prefer to always have a space after keywords and you should do the assignment on a new line.

File: modules/ffmpeg/producer/ffmpeg_producer.cpp [hunk 8]
```diff
@@ -608,27 +612,28 @@ struct ffmpeg_producer : public core::frame_producer_base
 void describe_producer(core::help_sink& sink, const core::help_repository& repo)
 {
 	sink.short_description(L"A producer for playing media files supported by FFmpeg.");
-	sink.syntax(L"[clip,url:string] {[loop:LOOP]} {SEEK [start:int]} {LENGTH [start:int]} {FILTER [filter:string]} {CHANNEL_LAYOUT [channel_layout:string]}");
+	sink.syntax(L"[clip,url:string] {[loop:LOOP]} {IN [in:int]} {OUT [out:int] | LENGTH [length:int]} {FILTER [filter:string]} {CHANNEL_LAYOUT [channel_layout:string]}");
```
Comment (HellGore) [inline modules/ffmpeg/producer/ffmpeg_producer.cpp]: `{IN [in:int]}` should be changed to `{IN,SEEK [in:int]}`

File: modules/ffmpeg/producer/ffmpeg_producer.cpp [hunk 9]
```diff
@@ -405,74 +408,77 @@ struct ffmpeg_producer : public core::frame_producer_base
...
-				input_.loop(boost::lexical_cast<bool>(value));
-			result = boost::lexical_cast<std::wstring>(input_.loop());
-		}
-		else if(boost::regex_match(param, what, seek_exp))
+		if(params.size())
 		{
-			auto value = boost::lexical_cast<int64_t>(what["VALUE"].str());
-			auto whence = what["WHENCE"].str();
-			auto total = file_nb_frames();
+			std::wstring cmd = params.at(0), value;
```
Comment (dimitry-ishenko) [inline modules/ffmpeg/producer/ffmpeg_producer.cpp]: Fixed

File: modules/ffmpeg/producer/ffmpeg_producer.cpp (hunk 8 above)
Comment (dimitry-ishenko) [inline modules/ffmpeg/producer/ffmpeg_producer.cpp]: Fixed

File: modules/ffmpeg/producer/ffmpeg_producer.cpp (hunk 7 above)
Comment (dimitry-ishenko) [inline modules/ffmpeg/producer/ffmpeg_producer.cpp]: I figured. I've fixed it everywhere and will resubmit in a minute

**Comment by dimitry-ishenko** [comment]
All fixed and resubmitted

**Comment by dimitry-ishenko** [comment]
@HellGore in case you've missed it, I had made changes based on your review and resubmitted the patch

**Comment by dimitry-ishenko** [comment]
@HellGore since you are on a roll, any chance of this being commited? :)

**Comment by jesperstarkar** [comment]
Oh, it's hard to keep up with all the great changes these days! 😄

**Comment by 5opr4ni** [comment]
Keep it rolling

**Comment by dimitry-ishenko** [comment]
Thank you

**Pull request merged**